src/layer/rnn.rs

```diff
@@ -25,26 +25,21 @@ impl Layer for RNN
                         , &ltex.weights[0]
                         , MatProp::NONE
                         , MatProp::NONE);
-
-    // uh_tm1 = htm1 U [bias added in h_t]
-    // handle case where time = 0
-    let uhtm1 = match ltex.recurrences.len(){
-      0 => {
-        let output_size = ltex.weights[1].dims()[0]; // is [M x M]
+    if current_unroll == 0 {
+        let output_size = ltex.weights[1].dims()[0];
         let init_h_dims = Dim4::new(&[inputs.dims()[0], output_size, 1, 1]);
-        ltex.recurrences.push(utils::constant(init_h_dims, inputs.get_type(), 0f32));
-        af::matmul(&utils::constant(init_h_dims, inputs.get_type(), 0f32)
-                   , &ltex.weights[1]
-                   , MatProp::NONE
-                   , MatProp::NONE)
-      },
-      _ => {
-        af::matmul(&ltex.recurrences[ltex.current_unroll]
+        if ltex.recurrences.len() == 0 {
+            ltex.recurrences.push(utils::constant(init_h_dims, inputs.get_type(), 0f32));
+        }
+        else {
+            ltex.recurrences[0] = utils::constant(init_h_dims, inputs.get_type(), 0f32);
+        }
+    }
+
+    let uhtm1 = af::matmul(&ltex.recurrences[current_unroll]
```

Comment:
You're right, no need to reinitialize h_0 since it doesn't change over time. However, we need to keep it in order to make the derivative later (not to handle a special case in the backward function)